qiskit/circuit/quantumcircuit.py

```diff
@@ -465,35 +465,70 @@ def reverse_bits(self) -> "QuantumCircuit":
         Examples:
 
             input:
-                 ┌───┐
-            q_0: ┤ H ├─────■──────
-                 └───┘┌────┴─────┐
-            q_1: ─────┤ RX(1.57) ├
-                      └──────────┘
+
+            .. parsed-literal::
+
+                     ┌───┐
+                a_0: ┤ H ├──■─────────────────
+                     └───┘┌─┴─┐
+                a_1: ─────┤ X ├──■────────────
+                          └───┘┌─┴─┐
+                a_2: ──────────┤ X ├──■───────
+                               └───┘┌─┴─┐
+                b_0: ───────────────┤ X ├──■──
+                                    └───┘┌─┴─┐
+                b_1: ────────────────────┤ X ├
+                                         └───┘
 
             output:
-                      ┌──────────┐
-            q_0: ─────┤ RX(1.57) ├
-                 ┌───┐└────┬─────┘
-            q_1: ┤ H ├─────■──────
-                 └───┘
+
+            .. parsed-literal::
+
+                                         ┌───┐
+                b_0: ────────────────────┤ X ├
+                                    ┌───┐└─┬─┘
+                b_1: ───────────────┤ X ├──■──
+                               ┌───┐└─┬─┘
+                a_0: ──────────┤ X ├──■───────
+                          ┌───┐└─┬─┘
+                a_1: ─────┤ X ├──■────────────
+                     ┌───┐└─┬─┘
+                a_2: ┤ H ├──■─────────────────
+                     └───┘
         """
         circ = QuantumCircuit(
-            *reversed(self.qregs),
-            *reversed(self.cregs),
             name=self.name,
             global_phase=self.global_phase,
         )
-        num_qubits = self.num_qubits
-        num_clbits = self.num_clbits
-        old_qubits = self.qubits
-        old_clbits = self.clbits
-        new_qubits = circ.qubits
-        new_clbits = circ.clbits
+        for bits, indices in [
+            (self.qubits, self._qubit_indices),
+            (self.clbits, self._clbit_indices),
+        ]:
+            i = len(bits) - 1
+            while i >= 0:
+                bit = bits[i]
+                _, regs = indices[bit]
+                if regs:
+                    reg, local_idx = regs[0]
+                    circ.add_register(reg)
+                    i -= reg.size
+                else:
+                    circ.add_bits([bit])
+                    i -= 1
```

Comment:
We also need to handle cases where registers overlap each other, i.e. the same bit is in multiple registers.  I think the cleanest solution to this whole function will be (completely untested, so I could be wrong on the logic):

1. create a circuit with no registers, and the same bit instances in reverse order:
    ```python
    out = QuantumCircuit(
        reversed(self.qubits),
        reversed(self.clbits),
        name=self.name,
        global_phase=self.global_phase,
    )
    ```
2. replay all the data from the old circuit into the new circuit _without changing the args_; because we reversed the bits, and the args all contain references to the bit instances, they'll automatically apply to the correct bits
3. go through each register in the old circuit, and make a new register that references the correct bits in the new circuit

I _think_ that's the correct logic, and it makes the correct separation between `Bit` and `Register`.

Reply:
Thanks for the suggestion. I followed these steps in the commit d0d8cc8.